iroh-net/src/discovery/local_node_discovery.rs

```diff
@@ -0,0 +1,250 @@
+                            ?discovered_node_id,
+                            ?peer_info,
+                            "LocalNodeDiscovery Message::Discovery"
+                        );
+                        let discovered_node_id = match PublicKey::from_str(&discovered_node_id) {
+                            Ok(node_id) => node_id,
+                            Err(e) => {
+                                tracing::warn!(
+                                    discovered_node_id,
+                                    "couldn't parse node_id from mdns discovery service: {e:?}"
+                                );
+                                continue;
+                            }
+                        };
+
+                        if discovered_node_id == node_id {
+                            continue;
+                        }
+
+                        if peer_info.is_expiry() {
+                            tracing::trace!(
+                                ?discovered_node_id,
+                                "removing node from LocalNodeDiscovery address book"
+                            );
+                            node_addrs.remove(&discovered_node_id);
+                            continue;
+                        }
+
+                        if let Some(sender) = senders.get(&discovered_node_id) {
+                            let item: DiscoveryItem = (&peer_info).into();
+                            tracing::trace!(?item, "sending DiscoveryItem");
+                            sender.send_async(Ok(item)).await.ok();
+                        }
+                        tracing::trace!(
+                            ?discovered_node_id,
+                            ?peer_info,
+                            "adding node to LocalNodeDiscovery address book"
+                        );
+                        node_addrs.insert(discovered_node_id, peer_info);
+                    }
+                    Message::SendAddrs((node_id, sender)) => {
+                        tracing::trace!(?node_id, "LocalNodeDiscovery Message::SendAddrs");
+                        if let Some(peer_info) = node_addrs.get(&node_id) {
+                            let item: DiscoveryItem = peer_info.into();
+                            tracing::debug!(?item, "sending DiscoveryItem");
+                            sender.send_async(Ok(item)).await.ok();
+                        }
+                        senders.insert(node_id, sender);
+                        let timeout_sender = task_sender.clone();
+                        timeouts.spawn(async move {
+                            tokio::time::sleep(DISCOVERY_DURATION).await;
+                            tracing::trace!(?node_id, "discovery timeout");
+                            timeout_sender
+                                .send_async(Message::Timeout(node_id))
+                                .await
+                                .ok();
+                        });
```

Comment:
I personally didn't want to add another `select` statement, since we seem to always have trouble with them. That was my logic for doing it this way. I can refactor tho, if folks have a differing opinion.